src/StepAdaptationController/include/WalkingControllers/StepAdaptationController/QPSolver.hpp

```diff
@@ -0,0 +1,139 @@
+/**
+ * @file StepAdaptationController.hpp
+ * @authors Milad Shafiee <milad.shafiee@iit.it>
+ * @copyright 2020 iCub Facility - Istituto Italiano di Tecnologia
+ *            Released under the terms of the LGPLv2.1 or later, see LGPL.TXT
+ * @date 2020
+ */
+
+#ifndef WALKING_CONTROLLERS_QP_SOLVER_H
+#define WALKING_CONTROLLERS_QP_SOLVER_H
+
+// std
+#include <deque>
+#include <memory>
+// iDynTree
+#include <iDynTree/Core/SparseMatrix.h>
+#include <iDynTree/Core/VectorDynSize.h>
+#include <iDynTree/Core/MatrixDynSize.h>
+
+// osqp-eigen
+#include <OsqpEigen/OsqpEigen.h>
+
+#include <qpOASES.hpp>
+#include <WalkingControllers/iDynTreeUtilities/Helper.h>
+
+namespace WalkingControllers
+{
+/**
+ * QPSolver class
+ */
+    class QPSolver
```

Comment:
This class has a very generic name ``QPSolver`` but the interface is very specific to the problem you are trying to solve and not a generic QP problem. I would suggest the following:
-  change the name of the class to something more specific, like ``StepAdaptationQPSolver``
- change the name of the methods. For example:
    - ``setHessianMatrix`` :arrow_right:  ``setCostFunctionWeights``
    - ``setConstraintsMatrix`` :arrow_right: ``setCurrentState`` (where you pass ``currentDcmPosition``, ``currentZmpPosition`` and ``remainingSingleSupportDuration``) + ``setConvexHull`` where you pass the allowable convex hull, both the matrix and the vector.
    - ``setGradientVector`` :arrow_right: ``setNominalValues``(where you pass ``zmpNominal``, ``dcmOffsetNominal``, ``sigmaNominal``, ``stepDuration`` and ``stepDurationTolerance``)
    - ``setBoundsVectorOfConstraints`` :arrow_right: ``setOmega`` where you specify only omega. All the other quantities are passed in the previous methods.

Reply:
>    - ``setConstraintsMatrix`` :arrow_right: ``setCurrentState`` (where you pass ``currentDcmPosition``, ``currentZmpPosition`` and ``remainingSingleSupportDuration``) + ``setConvexHull`` where you pass the allowable convex hull, both the matrix and the vector.

Why do you think it may cause a problem? Here we have a specific optimization problem for step adaptation that has a constraint matrix. Then here we set our specific constraint matrix.

Reply:
Also for other function name changing, inside the StepAdaptationQPSolver we set the hessian, gradient and any other required part related to the step adaptation QP.

Reply:
The problems here are two:
- the same input can be fed to two or more methods. Just to name one, ``zmpWeight`` needs to be set both to ``setHessianMatrix`` and to ``setGradientVector``. This is clearly not desirable as I could do a mistake and put two different ``zmpWeight`` and the result of the QP would be different.
- Readability. Suppose I do the following:
```
setHessianMatrix(w1, w2, 0.1);
```
One, reading the code, may wonder why a method that sets a hessian of a QP problem needs 3 parameters. In principle, you'll need to pass a matrix. Thus, it is always desirable that the name of the method reflects what it is needed for. In this case, we are "setting" something (hence the prefix ``set``). Setting what? From the outside, I would say **not** the hessian since I am passing something else. More specifically I am passing a bunch of weights used in the cost function. Hence, ``setCostFunctionWeights``. Then inside you can do whatever you want with that data and the user does not necessarily need to know what you are doing under the hood.

Reply:
In other words, the user of this class does not need to know how the underlying QP problem is built. In principle, you could do that when you call the ``solve`` method.

Reply:
Actually a user just uses stepAdaptationController class  that has been defined more or less similar in the way that you have suggested above! 
https://github.com/MiladShafiee/walking-controllers/blob/960f9cb580dfb990a666a7c0469239b4f229a1b6/src/StepAdaptationController/src/StepAdaptationController.cpp#L150

Then inside the solve method we call setHessianMatrix and other stuff: https://github.com/MiladShafiee/walking-controllers/blob/960f9cb580dfb990a666a7c0469239b4f229a1b6/src/StepAdaptationController/src/StepAdaptationController.cpp#L190

Reply:
The two problems [listed above](https://github.com/MiladShafiee/walking-controllers/pull/2#discussion_r388277193) remain though.

If this class is not supposed to be used by any other, you could simplify even more the interface, maybe having a single method where you pass a struct (or more, we discuss this) containing all the parameters that you need.

Reply:
> the same input can be fed to two or more methods. Just to name one, zmpWeight needs to be set both to setHessianMatrix and to setGradientVector. This is clearly not desirable as I could do a mistake and put two different zmpWeight and the result of the QP would be different.

> In other words, the user of this class does not need to know how the underlying QP problem is built. In principle, you could do that when you call the solve method.


I think this problem is solved because the user call only the following functions(that is inside the `stepAdaptationController` class:
`setCurrentDcmPosition` `setCurrentZmpPosition` `setNominalDcmOffset` and .....


> Thus, it is always desirable that the name of the method reflects what it is needed for. In this case, we are "setting" something (hence the prefix set). Setting what? From the outside, I would say not the hessian since I am passing something else.

For this problem for example we can change the name to `configureHessianMatrix` instead of `setHessianMAtrix` . What do you think?

Reply:
> I think this problem is solved because the user call only the following functions(that is inside the stepAdaptationController class:
setCurrentDcmPosition setCurrentZmpPosition setNominalDcmOffset and .....

I disagree. Indeed, the user of this class is ``stepAdaptationController``, hence the problem is still there. 

Anyways, I think there is a design problem here. Probably it is better to discuss F2F to reach an agreement on a suitable design.

Reply:
OK we can discuss F2F on Monday.